File: data-structure-and-algorithm/java/subsets.java
/**
 * Subsets
 * 
 * Given a set of distinct integers, S, return all possible subsets.
 * 
 * Note:
 * 
 * Elements in a subset must be in non-descending order.
 * The solution set must not contain duplicate subsets.
 *  
 * 
 * For example,
 * If S = [1,2,3], a solution is:
 * 
 * [
 *   [3],
 *   [1],
 *   [2],
 *   [1,2,3],
 *   [1,3],
 *   [2,3],
 *   [1,2],
 *   []
 * ]
 */

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        if (nums == null) {
            throw new NullPointerException();
        }

        Arrays.sort(nums);

        List<List<Integer>> result = new ArrayList<List<Integer>>();
        List<Integer> path = new ArrayList<>();
        subsetsHelper(nums, 0, path, result);
        return result;
    }

    private void subsetsHelper(int[] nums, int start, List<Integer> path, List<List<Integer>> result) {
        result.add(path);
        for (int i = start; i < nums.length; ++i) {
            path.add(nums[i]);
            subsetsHelper(nums, i + 1, path, result);
            path.remove(path.size() - 1);
        }
    }
}

File: data-structure-and-algorithm/typescript/leetcode/src/subsets.ts
/*
Given an integer array nums of unique elements, return all possible subsets (the power set).

The solution set must not contain duplicate subsets. Return the solution in any order.



Example 1:

Input: nums = [1,2,3]
Output: [[],[1],[2],[1,2],[3],[1,3],[2,3],[1,2,3]]
Example 2:

Input: nums = [0]
Output: [[],[0]]


Constraints:

1 <= nums.length <= 10
-10 <= nums[i] <= 10
All the numbers of nums are unique.
*/

function subsets(nums: number[]): number[][] {
  const result: number[][] = [];
  nums.sort((a, b) => a - b);
  const path: number[] = [];
  subsetsDfs(0);
  return result;

  function subsetsDfs(start: number): void {
    result.push([...path]);
    for (let i = start; i < nums.length; ++i) {
      path.push(nums[i]);
      subsetsDfs(i + 1);
      path.pop();
    }
  }
}
